google/cloud/security/scanner/audit/buckets_rules_engine.py

```diff
@@ -0,0 +1,255 @@
+
+        Args:
+            rules_file_path: file location of rules
+        """
+        super(BucketsRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self):
+        """Build BucketsRuleBook from the rules definition file."""
+        self.rule_book = BucketsRuleBook(self._load_rule_definitions())
+
+    # pylint: disable=arguments-differ
+    def find_policy_violations(self, buckets_acls,
+                               force_rebuild=False):
+        """Determine whether bucket acls violates rules."""
+        violations = itertools.chain()
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+        resource_rules = self.rule_book.get_resource_rules()
+
+        for rule in resource_rules:
+            violations = itertools.chain(violations,
+                                         rule.\
+                                         find_policy_violations(buckets_acls))
+        return violations
+
+    def add_rules(self, rules):
+        """Add rules to the rule book."""
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+class BucketsRuleBook(bre.BaseRuleBook):
+    """The RuleBook for bucket acls resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs: rule definitons
+        """
+        super(BucketsRuleBook, self).__init__()
+        self.resource_rules_map = {}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book"""
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def: A dictionary containing rule definition properties.
+            rule_index: The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+        """
+
+        resources = rule_def.get('resource')
+
+        for resource in resources:
+            resource_ids = resource.get('resource_ids')
+
+            if not resource_ids or len(resource_ids) < 1:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Missing resource ids in rule {}'.format(rule_index))
+
+            bucket = rule_def.get('bucket')
+            entity = rule_def.get('entity')
+            email = rule_def.get('email')
+            domain = rule_def.get('domain')
+            role = rule_def.get('role')
+
+            if (bucket is None) or (entity is None) or (email is None) or\
+               (domain is None) or (role is None):
+                LOGGER.error('Faulty rule: %s', rule_def.get('name'))
+                sys.exit()
```

Comment:
I think for now, since we are running single scanner pipelines at the moment, it's ok to call sys.exit(), but maybe a future TODO is to raise an error.